File: apps/apple/HealthMd/Shared/ClinicianReport/ClinicianReportModels.swift
import Foundation

/// Ephemeral report configuration. This is intentionally not Codable or persisted.
nonisolated struct ReportConfiguration: Equatable {
    var dateRange: ReportDateRange
    var selectedMetrics: Set<ReportMetric>
    var detailLevel: ReportDetailLevel
    var unitPreference: UnitPreference
    var displayName: String

    init(
        dateRange: ReportDateRange = .preset(.days30),
        selectedMetrics: Set<ReportMetric> = Set(ReportMetric.allCases),
        detailLevel: ReportDetailLevel = .summary,
        unitPreference: UnitPreference = .metric,
        displayName: String = ""
    ) {
        self.dateRange = dateRange
        self.selectedMetrics = selectedMetrics
        self.detailLevel = detailLevel
        self.unitPreference = unitPreference
        self.displayName = displayName
    }
}

nonisolated enum ReportDateRangePreset: String, CaseIterable, Identifiable {
    case days7
    case days30
    case days90
    case custom

    var id: String { rawValue }

    var dayCount: Int? {
        switch self {
        case .days7: return 7
        case .days30: return 30
        case .days90: return 90
        case .custom: return nil
        }
    }

    func title(using copy: ClinicianReportCopy) -> String {
        switch self {
        case .days7: return copy.string(.days_7)
        case .days30: return copy.string(.days_30)
        case .days90: return copy.string(.days_90)
        case .custom: return copy.string(.custom)
        }
    }

    var title: String { title(using: ClinicianReportCopy()) }
}

nonisolated struct ReportDateRange: Equatable {
    let startDate: Date
    let endDate: Date

    static func preset(
        _ preset: ReportDateRangePreset,
        today: Date = Date(),
        calendar: Calendar = .current
    ) -> ReportDateRange {
        precondition(preset != .custom)
        let todayStart = calendar.startOfDay(for: today)
        let start = calendar.date(byAdding: .day, value: -(preset.dayCount! - 1), to: todayStart)!
        return ReportDateRange(startDate: start, endDate: todayStart)
    }

    static func normalized(
        start: Date,
        end: Date,
        today: Date = Date(),
        calendar: Calendar = .current
    ) -> ReportDateRange {
        let todayStart = calendar.startOfDay(for: today)
        let first = min(calendar.startOfDay(for: start), todayStart)
        let last = min(calendar.startOfDay(for: end), todayStart)
        return ReportDateRange(startDate: min(first, last), endDate: max(first, last))
    }

    func normalized(today: Date = Date(), calendar: Calendar) -> ReportDateRange {
        .normalized(start: startDate, end: endDate, today: today, calendar: calendar)
    }

    func dates(calendar: Calendar) -> [Date] {
        var result: [Date] = []
        var cursor = calendar.startOfDay(for: startDate)
        let final = calendar.startOfDay(for: endDate)
        while cursor <= final {
            result.append(cursor)
            guard let next = calendar.date(byAdding: .day, value: 1, to: cursor) else { break }
            cursor = next
        }
        return result
    }

    func interval(calendar: Calendar) -> Range<Date> {
        let start = calendar.startOfDay(for: startDate)
        let end = calendar.date(byAdding: .day, value: 1, to: calendar.startOfDay(for: endDate))!
        return start..<end
    }

    func inclusiveDayCount(calendar: Calendar) -> Int {
        dates(calendar: calendar).count
    }
}

nonisolated enum ReportDetailLevel: String, CaseIterable, Identifiable {
    case summary
    case summaryAndReadings
    var id: String { rawValue }
}

nonisolated enum ReportMetric: String, CaseIterable, Identifiable, Hashable, Sendable {
    case bloodPressure
    case restingHeartRate
    case heartRate
    case weight
    case bloodGlucose
    case oxygenSaturation
    case respiratoryRate
    case bodyTemperature
    case sleepDuration
    case steps
    case workouts

    var id: String { rawValue }

    func displayName(using copy: ClinicianReportCopy) -> String {
        switch self {
        case .bloodPressure: return copy.string(.metric_blood_pressure)
        case .restingHeartRate: return copy.string(.metric_resting_heart_rate)
        case .heartRate: return copy.string(.metric_heart_rate)
        case .weight: return copy.string(.metric_weight)
        case .bloodGlucose: return copy.string(.metric_blood_glucose)
        case .oxygenSaturation: return copy.string(.metric_oxygen_saturation)
        case .respiratoryRate: return copy.string(.metric_respiratory_rate)
        case .bodyTemperature: return copy.string(.metric_body_temperature)
        case .sleepDuration: return copy.string(.metric_sleep_duration)
        case .steps: return copy.string(.metric_steps)
        case .workouts: return copy.string(.metric_workouts)
        }
    }

    var displayName: String { displayName(using: ClinicianReportCopy()) }
}

nonisolated struct ReportSource: Equatable {
    let label: String
    let isManualEntry: Bool

    init(label: String, isManualEntry: Bool = false) {
        self.label = label
        self.isManualEntry = isManualEntry
    }

    func displayLabel(using copy: ClinicianReportCopy) -> String {
        guard isManualEntry else { return label }
        let manualEntry = copy.string(.manual_entry)
        guard !label.localizedCaseInsensitiveContains(manualEntry) else { return label }
        return label.isEmpty ? manualEntry : copy.format(.manual_entry_source, label)
    }

    var displayLabel: String { displayLabel(using: ClinicianReportCopy()) }
}

nonisolated struct ScalarReportObservation: Equatable {
    let metric: ReportMetric
    let timestamp: Date
    let value: Double
    let stableID: String?
    let source: ReportSource?

    init(metric: ReportMetric, timestamp: Date, value: Double, stableID: String? = nil, source: ReportSource? = nil) {
        self.metric = metric
        self.timestamp = timestamp
        self.value = value
        self.stableID = stableID
        self.source = source
    }
}

nonisolated struct BloodPressureReportObservation: Equatable {
    let timestamp: Date
    let systolic: Double
    let diastolic: Double
    let stableID: String?
    let source: ReportSource?
}

nonisolated struct DailyReportValue: Equatable {
    let metric: ReportMetric
    let date: Date
    let value: Double
    let timestamp: Date?
    let stableID: String?
    let source: ReportSource?

    init(metric: ReportMetric, date: Date, value: Double, timestamp: Date? = nil, stableID: String? = nil, source: ReportSource? = nil) {
        self.metric = metric
        self.date = date
        self.value = value
        self.timestamp = timestamp
        self.stableID = stableID
        self.source = source
    }
}

nonisolated struct SleepReportObservation: Equatable {
    let date: Date
    let durationMinutes: Double
    let stableID: String?
    let source: ReportSource?
}

nonisolated struct WorkoutReportObservation: Equatable {
    let timestamp: Date
    let type: WorkoutType
    let durationMinutes: Double
    let stableID: String?
    let source: ReportSource?
}

nonisolated struct ClinicianReportInput: Equatable {
    let configuration: ReportConfiguration
    let calendar: Calendar
    let generatedAt: Date
    var scalarObservations: [ScalarReportObservation] = []
    var bloodPressureObservations: [BloodPressureReportObservation] = []
    var dailyValues: [DailyReportValue] = []
    var sleepObservations: [SleepReportObservation] = []
    var workoutObservations: [WorkoutReportObservation] = []
    var warnings: [String] = []
}

nonisolated struct ReportFact: Equatable, Sendable {
    let label: String
    let value: String
}

nonisolated struct ReportTable: Equatable, Sendable {
    let title: String
    let columns: [String]
    let rows: [[String]]
}

nonisolated struct MetricReportSummary: Equatable, Identifiable, Sendable {
    let metric: ReportMetric
    let facts: [ReportFact]
    let sources: [String]
    let coverageDisclosure: String?
    let noDataMessage: String?
    let table: ReportTable?
    var localizedTitle: String = ""
    var sourcesDisclosure: String? = nil
    var detailReadingsDescription: String? = nil
    var id: ReportMetric { metric }
}

nonisolated enum ReportCompleteness: Equatable, Sendable {
    case complete
    case partial
}

nonisolated struct ClinicianReportData: Equatable, Sendable {
    let title: String
    let displayName: String?
    let dateRangeLabel: String
    let generatedLabel: String
    let timeZoneLabel: String
    let sections: [MetricReportSummary]
    let warnings: [String]
    let completeness: ReportCompleteness
    let disclaimer: String
    let attribution: String
    let practiceLine: String?
    var languageTag: String = "en"
    /// Region captured with the report locale solely for native Letter/A4 selection.
    var paperRegionCode: String? = nil
    var pdfSubject: String = ""
    var pdfKeywords: [String] = []
    var metadataPeriodLabel: String = ""
    var metadataGeneratedLabel: String = ""
    var metadataTimeZoneLabel: String = ""
    var metadataPatientLabel: String = ""
    var availabilityNoteTitle: String = ""
    var aboutTitle: String = ""
    var pageFooterTemplate: String = ""
}

File: apps/android/app/src/main/java/com/healthmd/domain/clinicianreport/ClinicianReportModels.kt
package com.healthmd.domain.clinicianreport

import com.healthmd.domain.model.UnitPreference
import com.healthmd.domain.model.WorkoutType
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

/** Internal report configuration. It is intentionally not persisted or part of an export schema. */
data class ReportConfiguration(
    val dateRange: ReportDateRange = ReportDateRange.preset(ReportDateRangePreset.DAYS_30),
    val selectedMetrics: Set<ReportMetric> = ReportMetric.entries.toSet(),
    val detailLevel: ReportDetailLevel = ReportDetailLevel.SUMMARY,
    val unitPreference: UnitPreference = UnitPreference.METRIC,
    val displayName: String = "",
)

enum class ReportDateRangePreset(val dayCount: Long) {
    DAYS_7(7), DAYS_30(30), DAYS_90(90), CUSTOM(0);
}

data class ReportDateRange(val startDate: LocalDate, val endDate: LocalDate) {
    val inclusiveDayCount: Int get() = generateSequence(startDate) { current ->
        current.plusDays(1).takeIf { it <= endDate }
    }.count()

    fun dates(): List<LocalDate> = generateSequence(startDate) { current ->
        current.plusDays(1).takeIf { it <= endDate }
    }.toList()

    fun interval(zoneId: ZoneId): ReportInstantRange = ReportInstantRange(
        startInclusive = startDate.atStartOfDay(zoneId).toInstant(),
        endExclusive = endDate.plusDays(1).atStartOfDay(zoneId).toInstant(),
    )

    companion object {
        fun preset(preset: ReportDateRangePreset, today: LocalDate = LocalDate.now()): ReportDateRange {
            require(preset != ReportDateRangePreset.CUSTOM)
            return ReportDateRange(today.minusDays(preset.dayCount - 1), today)
        }

        fun normalized(start: LocalDate, end: LocalDate, today: LocalDate = LocalDate.now()): ReportDateRange {
            val clampedStart = minOf(start, today)
            val clampedEnd = minOf(end, today)
            return ReportDateRange(minOf(clampedStart, clampedEnd), maxOf(clampedStart, clampedEnd))
        }
    }
}

data class ReportInstantRange(val startInclusive: Instant, val endExclusive: Instant) {
    fun contains(instant: Instant): Boolean = instant >= startInclusive && instant < endExclusive
}

enum class ReportDetailLevel { SUMMARY, SUMMARY_AND_READINGS }

enum class ReportMetric {
    BLOOD_PRESSURE,
    RESTING_HEART_RATE,
    HEART_RATE,
    WEIGHT,
    BLOOD_GLUCOSE,
    OXYGEN_SATURATION,
    RESPIRATORY_RATE,
    BODY_TEMPERATURE,
    SLEEP_DURATION,
    STEPS,
    WORKOUTS,
}

data class ReportSource(val label: String, val isManualEntry: Boolean = false) {
    fun displayLabel(vocabulary: ClinicianReportVocabulary): String {
        if (!isManualEntry) return label
        val manual = vocabulary.text(ClinicianReportText.MANUAL_ENTRY)
        if (label.isBlank()) return manual
        return vocabulary.text(ClinicianReportText.MANUAL_ENTRY_SOURCE, label)
    }
}

data class ScalarReportObservation(
    val metric: ReportMetric,
    val timestamp: Instant,
    val value: Double,
    val stableId: String? = null,
    val source: ReportSource? = null,
)

data class BloodPressureReportObservation(
    val timestamp: Instant,
    val systolic: Double,
    val diastolic: Double,
    val stableId: String? = null,
    val source: ReportSource? = null,
)

data class DailyReportValue(
    val metric: ReportMetric,
    val date: LocalDate,
    val value: Double,
    val stableId: String? = null,
    val source: ReportSource? = null,
)

data class SleepReportObservation(
    val date: LocalDate,
    val durationMinutes: Double,
    val stableId: String? = null,
    val source: ReportSource? = null,
)

data class WorkoutReportObservation(
    val timestamp: Instant,
    val type: WorkoutType,
    val durationMinutes: Double,
    val stableId: String? = null,
    val source: ReportSource? = null,
)

sealed interface ClinicianReportWarning {
    data object ReadFailure : ClinicianReportWarning
    data class SourceFailure(val date: LocalDate) : ClinicianReportWarning
}

data class ClinicianReportInput(
    val configuration: ReportConfiguration,
    val zoneId: ZoneId,
    val generatedAt: Instant,
    val scalarObservations: List<ScalarReportObservation> = emptyList(),
    val bloodPressureObservations: List<BloodPressureReportObservation> = emptyList(),
    val dailyValues: List<DailyReportValue> = emptyList(),
    val sleepObservations: List<SleepReportObservation> = emptyList(),
    val workoutObservations: List<WorkoutReportObservation> = emptyList(),
    val warnings: List<ClinicianReportWarning> = emptyList(),
)

data class ReportFact(val label: String, val value: String)

data class ReportTable(
    val title: String,
    val columns: List<String>,
    val rows: List<List<String>>,
)

data class MetricReportSummary(
    val metric: ReportMetric,
    val facts: List<ReportFact>,
    val sources: List<String>,
    val coverageDisclosure: String?,
    val noDataMessage: String?,
    val table: ReportTable?,
    val localizedTitle: String,
    val sourcesDisclosure: String? = null,
    val detailReadingsDescription: String? = null,
)

enum class ReportCompleteness { COMPLETE, PARTIAL }

data class ClinicianReportData(
    val title: String,
    val displayName: String?,
    val dateRangeLabel: String,
    val generatedLabel: String,
    val timeZoneLabel: String,
    val sections: List<MetricReportSummary>,
    val warnings: List<String>,
    val completeness: ReportCompleteness,
    val disclaimer: String,
    val attribution: String,
    val practiceLine: String?,
    val languageTag: String,
    val paperRegionCode: String?,
    val isRtl: Boolean,
    val pdfSubject: String,
    val pdfKeywords: List<String>,
    val metadataPeriodLabel: String,
    val metadataGeneratedLabel: String,
    val metadataTimeZoneLabel: String,
    val metadataPatientLabel: String,
    val availabilityNoteTitle: String,
    val aboutTitle: String,
    val pageFooterTemplate: String,
)
